client_ext_norace_test.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2021-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !race
+// +build !race
+
+package connect_test
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	connect "connectrpc.com/connect"
+	"connectrpc.com/connect/internal/assert"
+	pingv1 "connectrpc.com/connect/internal/gen/connect/ping/v1"
+	"connectrpc.com/connect/internal/gen/connect/ping/v1/pingv1connect"
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+)
+
+func TestClientDeadlineHandling(t *testing.T) {
+	t.Parallel()
+
+	_, handler := pingv1connect.NewPingServiceHandler(pingServer{})
+	svr := httptest.NewUnstartedServer(http.HandlerFunc(func(respWriter http.ResponseWriter, req *http.Request) {
+		if req.Context().Err() != nil {
+			return
+		}
+		handler.ServeHTTP(respWriter, req)
+	}))
+	svr.EnableHTTP2 = true
+	svr.StartTLS()
+	t.Cleanup(svr.Close)
+
+	// This run creates a new connection for each RPC to verify that timeouts during dialing
+	// won't cause issues. This is historically easier to reproduce, so it uses a smaller
+	// duration, no concurrency, and fewer iterations. This is important because if we used
+	// a new connection for each RPC in the bigger test scenario below, we'd encounter other
+	// issues related to overwhelming the loopback interface and exhausting ephemeral ports.
+	t.Run("dial", func(t *testing.T) {
+		t.Parallel()
+		transport, ok := svr.Client().Transport.(*http.Transport)
+		if !assert.True(t, ok) {
+			t.FailNow()
+		}
+		testClientDeadlineBruteForceLoop(t,
+			5*time.Second, 5, 1,
+			func(ctx context.Context) (string, rpcErrors) {
+				httpClient := &http.Client{
+					Transport: transport.Clone(),
+				}
```

Comment:
Why the `transport.Clone()` and not use a new client? I think this is effectively the same:
```go
httpClient := svr.Client()
```

Reply:
Possibly. I'll look into it. It wasn't obvious when I was looking at it that this would create a new a transport (and new associated connection pool).

But I could definitely simplify a little. I originally had some concurrency in this test (the dial test) as well, and I wanted to make sure there was no chance of any of the tests re-using, since I'm trying to make sure every RPC has to re-dial. But I've since changed this to use no parallelism (one goroutine) because (1) it wasn't necessary as the dial error repro's fairly consistently and (2) it would run into issues with exhausting ephemeral ports. In fact, the latter could happen even with one goroutine (hence the addition of one millisecond sleep after trying to close idle connections, to make sure ephemeral ports are released).

Reply:
`svr.Client()` re-uses the same client over and over, so definitely don't want to do that. It's fine for the `read-write` cases to use that client (and they do), but if we mix clients (and thus mix transports), we can't be sure that we are actually dialing for each RPC.